### PR TITLE
ansible_hostname_f: a platform fact mimicking `hostname -f`

### PR DESCRIPTION
**Pull request by dexpl**
##### SUMMARY

There are some issues with ansible_fqdn platform fact (#38777, #9972, maybe others). I recently stuck into another one: if you have multiple PTR records with different names for one IP address, ansible_fqdn value is randomly set to one of those names. For example, [Mikrotik](http://www.mikrotik.com) built-in DNS server lets you have multiple names resolving to the same IP, but it gives you no option to choose which name to resolve that IP into via PTR record — it just creates as many PTRs for that IP as many names you give it. Nevertheless, `hostname -f` always returns the same, there [was](https://github.com/ansible/ansible/issues/9972#issuecomment-69468592) a hint about it.

Due to this, I propose a new fact, ansible_hostname_f, which value is populated the same way `hostname -f` works:
``
Technically: The FQDN is the name getaddrinfo(3) returns for the host name returned by gethostname(2).  The DNS domain name is the part after the first dot.
``
(from `man hostname`). I also propose ansible_hostname_d mimicking `hostname -d`.

In order not to break the backwards compatibility with existing setups, I don't replace any existing facts but add new ones instead.

In order to maintain cross-platform portability this implementation does not call hostname(1) binary but implements the same algo on python using the already-utilised socket module.

Please consider applying.
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

lib/ansible/module_utils/facts/system/platform.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by dexpl** [comment]
bot_status

**Comment by dexpl** [comment]
!ready_for_review
As far as I can see, failed tests are due to some issues with shippable itself and not related to this PR. Please correct me if I'm wrong and give a hint about my further actions.

File: lib/ansible/module_utils/facts/system/platform.py [hunk 1]
```diff
@@ -53,6 +53,11 @@ def collect(self, module=None, collected_facts=None):
 
         platform_facts['domain'] = '.'.join(platform_facts['fqdn'].split('.')[1:])
 
+        # this is what `hostname -f` returns
+        platform_facts['hostname_f'] = socket.getaddrinfo(socket.gethostname(), None, 0, socket.SOCK_DGRAM, 0, socket.AI_CANONNAME)[0][3]
```
Comment (bcoca) [inline lib/ansible/module_utils/facts/system/platform.py:57]: not really, using your code i get: `ansible_hostname_f": "localhost",`
but `hostname -f` gives me `erebos`.

Comment (dexpl) [reply lib/ansible/module_utils/facts/system/platform.py:57]: That's surprising. Could you describe your environment?

**Comment by samdoran** [comment]
Getting hostname resolution "right" is very tricky, mainly because "right" varies highly. Rather than adding a new fact, use one of the available mechanisms to set the fact how you see fit:

1. [Local facts](https://docs.ansible.com/ansible/latest/inventory/implicit_localhost.html?highlight=localhost#implicit-localhost)
2. Write a custom facts module and run that explicitly as a task
3. As of Ansible 2.8.0, add your custom facts module to [facts_modules](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#facts-modules) so it will be run automatically during fact gathering.

To discuss this further, please reach out on IRC or the mailing list:

   * IRC: #ansible-dev on irc.freenode.net   
   * mailing list: https://groups.google.com/forum/#!forum/ansible-devel

**Pull request closed without merging**